subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/OutputChangeDetector.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.history.changes;
+
+import org.gradle.internal.RelativePathSupplier;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.fingerprint.FileSystemLocationFingerprint;
+import org.gradle.internal.fingerprint.impl.DefaultFileSystemLocationFingerprint;
+import org.gradle.internal.snapshot.FileSystemLocationSnapshot;
+import org.gradle.internal.snapshot.FileSystemSnapshot;
+import org.gradle.internal.snapshot.RelativePathTracker;
+import org.gradle.internal.snapshot.SnapshotVisitResult;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import static org.gradle.internal.execution.history.changes.AbstractFingerprintCompareStrategy.FINGERPRINT_CHANGE_FACTORY;
+
+public class OutputChangeDetector {
+    private final CompareStrategy.ChangeDetector<FileSystemLocationFingerprint> delegate;
+
+    public OutputChangeDetector(CompareStrategy.ChangeDetector<FileSystemLocationFingerprint> delegate) {
+        this.delegate = delegate;
+    }
+
+    public boolean visitChangesSince(FileSystemSnapshot previous, FileSystemSnapshot current, String propertyTitle, ChangeVisitor visitor) {
+        if (previous == current) {
+            return true;
+        }
+        if (previous == FileSystemSnapshot.EMPTY) {
+            return visitAllChanges(current, visitor, (relativePath, snapshot) -> FINGERPRINT_CHANGE_FACTORY.added(
+                snapshot.getAbsolutePath(),
+                propertyTitle,
+                createFingerprint(relativePath, snapshot)
+            ));
+        } else if (current == FileSystemSnapshot.EMPTY) {
+            return visitAllChanges(previous, visitor, (relativePath, snapshot) -> FINGERPRINT_CHANGE_FACTORY.removed(
+                snapshot.getAbsolutePath(),
+                propertyTitle,
+                createFingerprint(relativePath, snapshot)
+            ));
+        } else {
+            FileSystemLocationSnapshot previousSnapshot = (FileSystemLocationSnapshot) previous;
+            FileSystemLocationSnapshot currentSnapshot = (FileSystemLocationSnapshot) current;
```

Comment:
It would be great if we could eliminate the need for this cast. I think I'll be able to do that as part of the "resolve task properties only once" work I'm doing now, but maybe there's something simple we can do in the context of this PR already?

Reply:
I took a look, though it is not straightforward. I could try to do this in a follow up PR.